tests/gdx-tests/src/com/badlogic/gdx/tests/g3d/shadows/system/BaseShadowSystem.java

```diff
@@ -0,0 +1,468 @@
+	/** Analyzer of near and far for spot and point lights */
+	protected NearFarAnalyzer nearFarAnalyzer;
+	/** Allocator which choose where to render shadow map in texture */
+	protected ShadowMapAllocator allocator;
+	/** Analyzer which compute how to create the camera for directional light */
+	protected DirectionalAnalyzer directionalAnalyzer;
+	/** Filter that choose if light must be rendered */
+	protected LightFilter lightFilter;
+	/** Framebuffer used to render all the depth maps */
+	protected FrameBuffer[] frameBuffers;
+	/** Current pass in the depth process */
+	protected int currentPass = -1;
+	/** Iterators for cameras */
+	protected Entries<SpotLight, LightProperties> spotCameraIterator;
+	protected Entries<DirectionalLight, LightProperties> dirCameraIterator;
+	protected Entries<PointLight, PointLightProperties> pointCameraIterator;
+	/** Current side in the point light cubemap */
+	protected int currentPointSide;
+	protected PointLightProperties currentPointProperties;
+	/** Tmp variables */
+	protected Vector3 tmpV3 = new Vector3();
+	protected Vector2 tmpV2 = new Vector2();
+	/** Shader providers used by this system */
+	protected ShaderProvider[] passShaderProviders;
+	protected ShaderProvider mainShaderProvider;
+	/** Current light and properties during shadowmap generation */
+	protected LightProperties currentLightProperties;
+	protected BaseLight currentLight;
+
+	/** Construct the system with the needed params.
+	 * @param nearFarAnalyzer Analyzer of near and far
+	 * @param allocator Allocator of shadow maps
+	 * @param directionalAnalyzer Analyze directional light to create orthographic camera
+	 * @param lightFilter Filter light to render */
+	public BaseShadowSystem (NearFarAnalyzer nearFarAnalyzer, ShadowMapAllocator allocator,
+		DirectionalAnalyzer directionalAnalyzer, LightFilter lightFilter) {
+		this.nearFarAnalyzer = nearFarAnalyzer;
+		this.allocator = allocator;
+		this.directionalAnalyzer = directionalAnalyzer;
+		this.lightFilter = lightFilter;
+		frameBuffers = new FrameBuffer[getPassQuantity()];
+		passShaderProviders = new ShaderProvider[getPassQuantity()];
+		init();
+	}
+
+	/** Construct the system with default values */
+	public BaseShadowSystem () {
+		this(new AABBNearFarAnalyzer(), new FixedShadowMapAllocator(FixedShadowMapAllocator.QUALITY_MED,
+			FixedShadowMapAllocator.QUANTITY_MAP_MED), new BoundingSphereDirectionalAnalyzer(), new FrustumLightFilter());
+	}
+
+	/** Initialize framebuffers and shader providers. You should call super.init() in subclass. */
+	protected void init () {
+		frameBuffers[PASS_1] = new FrameBuffer(Pixmap.Format.RGBA8888, allocator.getWidth(), allocator.getHeight(), true);
+	};
+
+	/** getPassQuantity should return at leat one. */
+	@Override
+	public abstract int getPassQuantity ();
+
+	@Override
+	public ShaderProvider getPassShaderProvider (int n) {
+		if (n >= passShaderProviders.length)
```

Comment:
I dont think that this check is needed here (otherwise it will throw an index out of bounds exception, which says the same). https://github.com/libgdx/libgdx/wiki/Contributing#performance-considerations